KAFKA-15356: Generate and persist directory IDs

soarez opened the pull request:
Part of [KIP-858](https://cwiki.apache.org/confluence/display/KAFKA/KIP-858%3A+Handle+JBOD+broker+disk+failure+in+KRaft).

* Generate a directory ID when formatting storage
* Generate a directory ID if missing in meta.properties at startup

https://issues.apache.org/jira/browse/KAFKA-15356

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soarez committed:
KAFKA-15356: Generate and persist directory IDs

https://issues.apache.org/jira/browse/KAFKA-15356

soarez commented:
This PR is marked as draft because the KIP hasn't yet been accepted but it is ready for a review. 

@rondagostino could you have a look at this one too?

pprovenzano commented:
LGTM!

rondagostino commented on clients/src/main/java/org/apache/kafka/common/Uuid.java:
```diff
@@ -27,16 +30,33 @@
+     */
+    public static final Uuid OFFLINE_DIR = ONE_UUID;
+
+    private static final Set<Uuid> RESERVED = new HashSet<>(Arrays.asList(METADATA_TOPIC_ID, ZERO_UUID, ONE_UUID, UNKNOWN_DIR, OFFLINE_DIR));
```
wrap with `Collections.unmodifiableSet`.  Also might be useful to make it public so we can use it in tests.  For example, in StorageToolTest, we could say `assertFalse(Uuid.RESERVED.contains(directoryId))`.

rondagostino commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -261,6 +262,40 @@ class LogManager(logDirs: Seq[File],
+  private def directoryIds(dirs: Seq[File]): Map[String, Uuid] = {
+    dirs.flatMap { dir =>
+      try {
+        val metadataCheckpoint = new BrokerMetadataCheckpoint(new File(dir, "meta.properties"))
```
Maybe we can move `  val brokerMetaPropsFile = "meta.properties"` from `class KafkaServer` to `object KafkaServer` and refer to that? This string value is sprinkled throughout the code base, and we should clean them up by refactoring out a reusable constant -- no time like the present.

rondagostino commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1343,6 +1378,8 @@ class LogManager(logDirs: Seq[File],
   }
 
+  def directoryIds: Set[Uuid] = dirIds.values.toSet
+
```
`dirIds` is never mutated, so maybe we can calculate this right after we calculate `dirIds` and just return what we calculated then as opposed to  creating a new Set every time?

rondagostino commented on core/src/main/scala/kafka/server/BrokerMetadataCheckpoint.scala:
```diff
@@ -63,6 +64,18 @@ class RawMetaProperties(val props: Properties = new Properties()) {
+      Option(props.getProperty(DirectoryIdKey))
+    } else {
+      None
+    }
```
`Option(props.getProperty(DirectoryIdKey))` for consistency with how it is done for the cluster ID?

rondagostino commented on core/src/main/scala/kafka/server/BrokerMetadataCheckpoint.scala:
```diff
@@ -71,13 +84,6 @@ class RawMetaProperties(val props: Properties = new Properties()) {
     props.setProperty(VersionKey, ver.toString)
   }
 
-  def requireVersion(expectedVersion: Int): Unit = {
-    if (version != expectedVersion) {
-      throw new RuntimeException(s"Expected version $expectedVersion, but got "+
-        s"version $version")
-    }
-  }
-
```
Agree can remove since unused.

rondagostino commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -431,7 +431,9 @@ object StorageTool extends Logging {
-      checkpoint.write(metaProperties.toProperties)
+      val rawProps = new RawMetaProperties(metaProperties.toProperties)
+      rawProps.directoryId = Uuid.randomUuid().toString
+      checkpoint.write(rawProps.props)
```
It is a bit convoluted to invoke `new RawMetaProperties(metaProperties.toProperties)` since that essentially creates 2 instances of `RawMetaProperties`.  Maybe we can do instead `checkpoint.write(metaProperties.toPropertiesWithDirectoryId(Uuid.randomUuid().toString))` by adding the new method `toPropertiesWithDirectoryId()`?

rondagostino commented on core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:
```diff
@@ -361,4 +362,24 @@ Found problem:
+      val properties = new BrokerMetadataCheckpoint(metaPropertiesFile).read().get
+      assertTrue(properties.containsKey("directory.id"))
+      val directoryId = Uuid.fromString(properties.getProperty("directory.id"))
+      assertNotEquals(Uuid.UNKNOWN_DIR, directoryId)
+      assertNotEquals(Uuid.OFFLINE_DIR, directoryId)
```
Maybe make Uuid.RESERVED public and invoke !contains() instead?

rondagostino reviewed (commented):
Looks good. Made a few comments/suggestions, none of which are mandatory, but let me know what you think.  Thanks for the PR!

soarez committed:
Refactor meta.properties constant

soarez committed:
Review suggestions

soarez committed:
Reserve 100 UUIDs & add SELECTED_DIR

soarez commented:
Thank you for the reviews @rondagostino and @pprovenzano.

I've addressed the suggestions, and also made the changes discussed earlier:

* Introduce a third reserved UUID
* Reserve the first 100 UUIDs

PTAL and LMK if everything still looks OK.

soarez committed:
Missed "meta.properties" reference

rondagostino commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -117,6 +117,8 @@ class LogManager(logDirs: Seq[File],
 
   private val dirLocks = lockLogDirs(liveLogDirs)
+  private val dirIds = directoryIds(liveLogDirs)
+  val directoryIds: Set[Uuid] = dirIds.values.toSet
```
Seems this is only used in the test, so can we do this instead?
```
  // visible for testing
  private[log] val directoryIds: Set[Uuid] = dirIds.values.toSet
```

soarez replied:
Thanks. Updated.

rondagostino reviewed (commented):
LGTM -- thanks for all the adjustments based on the prior review.  There's just one issue about visibility.  Can you take a look at that?  I can likely approve/merge after your comment and/or change based on that.

soarez committed:
Adjust visibility of directoryIds field

rondagostino reviewed (approved):
LGTM -- thanks for the changes!

rondagostino commented:
Test failures are unrelated. Merging to trunk.
